reactor-core/src/main/java/reactor/core/publisher/FluxWindowTimeout.java

```diff
@@ -26,52 +26,1590 @@
+					previousState = expectedState | HAS_WORK_IN_PROGRESS;
+
+					if (isCancelled(expectedState)) {
+						return;
+					}
+
+					if (isTerminated(expectedState)) {
+						final Throwable e = this.error;
+						if (e != null) {
+							this.actual.onError(e);
+						}
+						else {
+							this.actual.onComplete();
+						}
+						return;
+					}
+
+
+					if (!hasWorkInProgress(expectedState)) {
+						return;
+					}
+
+					continue;
+				}
+
+				if (n > 0) {
+					// here we are if we entered from the request() method
+					if (hasUnsentWindow) {
+						final InnerWindow<T> currentUnsentWindow = this.window;
+
+						// Delivers current unsent window
+						this.actual.onNext(currentUnsentWindow);
+
+						if (n != Long.MAX_VALUE) {
+							n = REQUESTED.decrementAndGet(this);
+						}
+
+						// Marks as sent current unsent window. Also, delivers
+						// onComplete to the window subscriber if it was not delivered
+						final long previousInnerWindowState = currentUnsentWindow.sendSent();
+
+						if (isTerminated(expectedState)) {
+							Throwable e = this.error;
+							if (e != null) {
+								if (!isTerminated(previousInnerWindowState)) {
+									currentUnsentWindow.sendError(e);
+								}
+								this.actual.onError(e);
+							}
+							else {
+								if (!isTerminated(previousInnerWindowState)) {
+									currentUnsentWindow.sendComplete();
+								}
+								this.actual.onComplete();
+							}
+							return;
+						}
+
+						// we should create next window if we see that
+						// currentUnsentWindow is terminated and nextWindowIndex is
+						// greater than currentWindowIndex
+						if (nextWindowIndex > activeWindowIndex && (InnerWindow.isTimeout(previousInnerWindowState) || InnerWindow.isTerminated(previousInnerWindowState))) {
+							final boolean shouldBeUnsent = n == 0;
+							final InnerWindow<T> nextWindow =
+									new InnerWindow<>(this.maxSize, this,
+											nextWindowIndex, shouldBeUnsent);
+
+							this.window = nextWindow;
+
+							if (!shouldBeUnsent) {
+								this.actual.onNext(nextWindow);
+
+								if (n != Long.MAX_VALUE) {
+									REQUESTED.decrementAndGet(this);
+								}
+							}
+
+							previousState = commitWork(this, expectedState, shouldBeUnsent);
+							expectedState =
+									(((previousState &~ACTIVE_WINDOW_INDEX_MASK) &~ HAS_UNSENT_WINDOW) ^ (expectedState == previousState ? HAS_WORK_IN_PROGRESS : 0)) |
+									incrementActiveWindowIndex(previousState) |
+									(shouldBeUnsent ? HAS_UNSENT_WINDOW : 0);
+							// we need to put unsent flag here because we check at the
+							// beginning of the loop the presence for unsent flag from
+							// the previousState
+							previousState = (previousState &~ HAS_UNSENT_WINDOW) | (shouldBeUnsent ? HAS_UNSENT_WINDOW : 0);
+
+							if (isCancelled(expectedState)) {
+								nextWindow.sendCancel();
+								if (shouldBeUnsent) {
+									Operators.onDiscard(nextWindow, this.actual.currentContext());
```

Comment:
One thing that came to my mind regarding which `currentContext` to use. There are two options:
 1. `this.actual.currentContext()` (only available here at the coordinator's level)
 2. `nextWindow.actual.currentContext()` (available in both cases)

I think that for consistency and predictability it would be better to always use the window's reachable context for the purpose of discarding the queued elements.

That said, there is an interesting aspect to inner Flux in a window operator: unlike in eg. `flatMap`, the subscription to the window can be a few steps down the line. Operators in the main sequence in between might define some `contextWrite` operation that isn't visible to the window as a result.

```java
myFlux.window(/*...*/)
    .doOnDiscard(e -> System.out.println("DISCARD A " + e")) // not invoked
    .concatMap(window -> window
        .doOnDiscard(e -> System.out.println("DISCARD from window: " + e"))
        .map(e -> e.toString())
    );
```


It would probably be worth it to document that in a discard support javadoc paragraph on the `Flux` API.